src/contracts/core/AVSDirectory.sol

```diff
@@ -48,21 +56,200 @@ contract AVSDirectory is
+     *  @notice Called by AVSs to add an operator to an operator set.
+     *
+     *  @param operator The address of the operator to be added to the operator set.
+     *  @param operatorSetIds The IDs of the operator sets.
+     *  @param operatorSignature The signature of the operator on their intent to register.
+     *
+     *  @dev msg.sender is used as the AVS.
+     *  @dev The operator must not have a pending deregistration from the operator set.
+     *  @dev If this is the first operator set in the AVS that the operator is
+     *  registering for, a OperatorAVSRegistrationStatusUpdated event is emitted with
+     *  a REGISTERED status.
      */
-    function registerOperatorToAVS(
+    function registerOperatorToOperatorSets(
         address operator,
+        uint32[] calldata operatorSetIds,
         ISignatureUtils.SignatureWithSaltAndExpiry memory operatorSignature
     ) external onlyWhenNotPaused(PAUSED_OPERATOR_REGISTER_DEREGISTER_TO_AVS) {
+        // Assert `operator` is actually an operator.
+        require(
+            delegation.isOperator(operator),
+            "AVSDirectory.registerOperatorToOperatorSets: operator not registered to EigenLayer yet"
+        );
+        // Assert operator's signature has not expired.
+        require(
+            operatorSignature.expiry >= block.timestamp,
+            "AVSDirectory.registerOperatorToOperatorSets: operator signature expired"
+        );
+        // Assert operator's signature `salt` has not already been spent.
+        require(
+            !operatorSaltIsSpent[operator][operatorSignature.salt],
+            "AVSDirectory.registerOperatorToOperatorSets: salt already spent"
+        );
+
+        if (operatorSignature.signature.length != 0) {
+            // Assert signature provided by `operator` is valid.
+            EIP1271SignatureUtils.checkSignature_EIP1271(
+                operator,
+                calculateOperatorSetRegistrationDigestHash({
+                    avs: msg.sender,
+                    operatorSetIds: operatorSetIds,
+                    salt: operatorSignature.salt,
+                    expiry: operatorSignature.expiry
+                }),
+                operatorSignature.signature
+            );
+        }
+
+        // Mutate `operatorSaltIsSpent` to `true` to prevent future respending.
+        operatorSaltIsSpent[operator][operatorSignature.salt] = true;
+
+        // Register `operator` if not already registered.
+        if (avsOperatorStatus[msg.sender][operator] != OperatorAVSRegistrationStatus.REGISTERED) {
+            avsOperatorStatus[msg.sender][operator] = OperatorAVSRegistrationStatus.REGISTERED;
+            emit OperatorAVSRegistrationStatusUpdated(operator, msg.sender, OperatorAVSRegistrationStatus.REGISTERED);
+        }
+
+        // Mutate calling AVS to operator set AVS status, preventing further legacy registrations.
+        if (!isOperatorSetAVS[msg.sender]) isOperatorSetAVS[msg.sender] = true;
```

Comment:
I'm not a fan of this - I think this opens the door to footgun accidentally making your AVS operator sets only. 

If we're committed to supporting non-operator set AVSs, we need an explicit method to upgrade like `becomeOperatorSetAVS`

Reply:
I'm not sure how we'd ensure that new AVSs are operator set AVSs by default though....